src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/SourceGenerators/SourceGeneratorProjectItem.cs

```diff
@@ -47,7 +48,21 @@ public override Stream Read()
             => throw new NotSupportedException("This API should not be invoked. We should instead be relying on " +
                 "the RazorSourceDocument associated with this item instead.");
 
-        public bool Equals(SourceGeneratorProjectItem? other) => other is not null && AdditionalText == other.AdditionalText;
+        public bool Equals(SourceGeneratorProjectItem? other)
+        {
+            if (ReferenceEquals(AdditionalText, other?.AdditionalText))
+            {
+                return true;
+            }
+
+            // In the compiler server when the generator driver cache is enabled the
+            // additional files are always different instances even if their content is the same.
+            // It's technically possible for these hashes to collide, but other things would
+            // also break in those cases, so for now we're okay with this.
+            var thisHash = AdditionalText.GetText()?.GetContentHash() ?? [];
+            var otherHash = other?.AdditionalText.GetText()?.GetContentHash() ?? [];
+            return Enumerable.SequenceEqual(thisHash, otherHash);
```

Comment:
I believe that `thisHash` and `otherHash` should be `ImmutableArray<byte>`, correct? If so, could you use the `SequenceEqual` extension method from `ImmutableArrayExtensions`? I believe that `Enumerable.SequenceEqual(...)` will box them.

```suggestion
            return thisHash.SequenceEqual(otherHash);
```